tenant/pkg/webhook/default_server/tenantnamespace/validating/tenantnamespace_create_update_handler.go

```diff
@@ -36,19 +40,45 @@ func init() {
 
 // TenantNamespaceCreateUpdateHandler handles TenantNamespace
 type TenantNamespaceCreateUpdateHandler struct {
-	// To use the client, you need to do the following:
-	// - uncomment it
-	// - import sigs.k8s.io/controller-runtime/pkg/client
-	// - uncomment the InjectClient method at the bottom of this file.
-	// Client  client.Client
+	Client client.Client
 
 	// Decoder decodes objects
 	Decoder types.Decoder
 }
 
-func (h *TenantNamespaceCreateUpdateHandler) validatingTenantNamespaceFn(ctx context.Context, obj *tenancyv1alpha1.TenantNamespace) (bool, string, error) {
-	// TODO(user): implement your admission logic
-	return true, "allowed to be admitted", nil
+func (h *TenantNamespaceCreateUpdateHandler) validateTenantNamespaceUpdate(obj *tenancyv1alpha1.TenantNamespace, oldobj *tenancyv1alpha1.TenantNamespace) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMetaUpdate(&obj.ObjectMeta, &oldobj.ObjectMeta, field.NewPath("metadata"))
+	if obj.Spec.Name != oldobj.Spec.Name {
+		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec").Child("name"), "update to name field in spec is forbiddern"))
+	}
+	return allErrs
+}
+
+func validateTenantNamespaceName(name string, prefix bool) []string {
+	// We don't have name requirement for now
+	return nil
+}
+func (h *TenantNamespaceCreateUpdateHandler) validateTenantNamespaceCreate(obj *tenancyv1alpha1.TenantNamespace) field.ErrorList {
+	path := field.NewPath("metadata")
+	allErrs := apivalidation.ValidateObjectMeta(&obj.ObjectMeta, true, validateTenantNamespaceName, path)
+
+	// Fetch tenant list
+	tenantList := &tenancyv1alpha1.TenantList{}
+	err := h.Client.List(context.TODO(), &client.ListOptions{}, tenantList)
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(path.Child("Namespace"), obj.Namespace, "cannot validate namespace because cannot get tenant list"))
```

Comment:
Is there a way we can separate user errors from system errors? This could be a 500 or 503, and it could be very useful to see that in the ultimate response rather than a blank "something went wrong" message.

Perhaps this function could return (field.ErrorList, error); if the `error` is non-nil, we can process it specially? Sadly, it seems like `controller-runtime` doesn't give us great primitives for this (eg if `errors.IsNotFound(err)` returns true, there's no way to automatically turn that into a `404` status code) but at least we could warn the user that something funny's going on. You already do this in the Update case by returning `http.StatusInternalServerError` which would probably work well enough here too.

Reply:
done